input-events/input-events-typing.html

```diff
@@ -1,82 +1,311 @@
+
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        await test_driver.send_keys(document.getElementById('rich'), "\uE006"); // Return
+
+        assert_equals(inputEventsLog.length, 2);
+        const beforeInputEvent = inputEventsLog[0];
+        const inputEvent = inputEventsLog[1];
+        assert_equals(beforeInputEvent.type, 'beforeinput');
+        assert_equals(inputEvent.type, 'input');
+        assert_equals(beforeInputEvent.inputType, 'insertParagraph');
+        assert_equals(inputEvent.inputType, 'insertParagraph');
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing RETURN');
+
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        await new test_driver.Actions()
+            .keyDown('\uE008') // Shift
+            .keyDown('\uE006') // Enter
+            .keyUp('\uE006')
+            .keyUp('\uE008')
+            .send();
+
+        assert_equals(inputEventsLog.length, 2);
+        const beforeInputEvent = inputEventsLog[0];
+        const inputEvent = inputEventsLog[1];
+        assert_equals(beforeInputEvent.type, 'beforeinput');
+        assert_equals(inputEvent.type, 'input');
+        assert_equals(beforeInputEvent.inputType, 'insertLineBreak');
+        assert_equals(inputEvent.inputType, 'insertLineBreak');
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing Shift+RETURN');
+
+    promise_test(async () => {
+        resetEditor('<p>Preexisting <i id="caret">c</i>ontent</p>');
+        await test_driver.click(document.querySelector('#caret'));
+        await test_driver.send_keys(document.querySelector('#caret'), "\uE017"); // Delete
+
+        assert_equals(inputEventsLog.length, 2);
+        const beforeInputEvent = inputEventsLog[0];
+        const inputEvent = inputEventsLog[1];
+        assert_equals(beforeInputEvent.type, 'beforeinput');
+        assert_equals(inputEvent.type, 'input');
+        assert_equals(beforeInputEvent.inputType, 'deleteContentForward');
+        assert_equals(inputEvent.inputType, 'deleteContentForward');
+        assert_equals(beforeInputEvent.data, inputEvent.data);
+        resetEditor();
+        return Promise.resolve();
+    }, 'It triggers beforeinput and input events on typing DELETE with pre-existing content');
+
+    promise_test(async () => {
+        resetEditor();
+        document.getElementById('rich').focus();
+        await test_driver.send_keys(document.getElementById('rich'), "\uE017"); // Delete
+        assert_equals(inputEventsLog.length, 2);
+        const beforeInputEvent = inputEventsLog[0];
+        const inputEvent = inputEventsLog[1];
```

Comment:
If you want to get a bit fancier: 

```suggestion
        const [inputEvent, beforeInputEvent] = inputEventsLog;
```

Will pull out the right things for you.

Reply:
Done